interfaces/cython/cantera/test/test_reaction.py

```diff
@@ -134,108 +134,293 @@ def setUpClass(cls):
+        self.check_rate(rate)
+
+    def test_negative_A(self):
+        # test reaction rate property
+        rate = self.from_parts()
+        self.assertFalse(rate.allow_negative_pre_exponential_factor)
+        rate.allow_negative_pre_exponential_factor = True
+        self.assertTrue(rate.allow_negative_pre_exponential_factor)
+
+    def test_additional_parts(self):
+        # test reaction rate property
+        rate = self.from_parts()
+        rr = self.gas.reaction(self._index).rate
+        rr.delta_enthalpy = rate.delta_enthalpy
+        self.assertNear(rate.activation_energy, rr.activation_energy)
+
+
+class FalloffRateTests(ReactionRateTests):
+    # test Falloff rate expressions
+    _n_data = [0]
+
+    @classmethod
+    def setUpClass(cls):
+        ReactionRateTests.setUpClass()
+        param = cls._input["low-P-rate-constant"]
+        cls._parts["low"] = ct.Arrhenius(param["A"], param["b"], param["Ea"])
+        param = cls._input["high-P-rate-constant"]
+        cls._parts["high"] = ct.Arrhenius(param["A"], param["b"], param["Ea"])
+
+    def update(self, rate):
+        self.gas.forward_rate_constants # force update of internal states
+        concm = self.gas.third_body_concentrations
+        rate.third_body_concentration = concm[self._index]
+        return rate
+
+    def test_data(self):
+        rate = self.from_parts()
+        for n in self._n_data:
+            rate.data = np.random.rand(n)
+
+    def test_third_body(self):
+        concm = self.gas.third_body_concentrations
+        self.assertIsFinite(concm[self._index])
+
+
+class TestLindemannRate(FalloffRateTests, utilities.CanteraTest):
+    # test Lindemann rate expressions
+
+    _cls = ct.LindemannRate
+    _type = "Lindemann"
+    _index = 7
+    _parts = {
+        "data": [],
+        }
+    _input = {
+        "type": "falloff",
+        "low-P-rate-constant": {"A": 2.3e+12, "b": -0.9, "Ea": -7112800.0},
+        "high-P-rate-constant": {"A": 7.4e+10, "b": -0.37, "Ea": 0.0},
+        }
+    _yaml = """
+        type: falloff
+        low-P-rate-constant: {A: 2.3e+12, b: -0.9, Ea: -7112800.0}
+        high-P-rate-constant: {A: 7.4e+10, b: -0.37, Ea: 0.0}
+        """
+
+
+class TestTroeRate(FalloffRateTests, utilities.CanteraTest):
+    # test Troe rate expressions
+
+    _cls = ct.TroeRate
+    _type = "Troe"
+    _index = 2
+    _parts = {"data": [0.7346, 94.0, 1756.0, 5182.0]}
+    _input = {
+        "type": "falloff",
+        "low-P-rate-constant": {"A": 2.3e+12, "b": -0.9, "Ea": -7112800.0},
+        "high-P-rate-constant": {"A": 7.4e+10, "b": -0.37, "Ea": 0.0},
+        "Troe": {"A": 0.7346, "T3": 94.0, "T1": 1756.0, "T2": 5182.0},
+        }
+    _yaml = """
+        type: falloff
+        low-P-rate-constant: {A: 2.3e+12, b: -0.9, Ea: -7112800.0}
+        high-P-rate-constant: {A: 7.4e+10, b: -0.37, Ea: 0.0}
+        Troe: {A: 0.7346, T3: 94.0, T1: 1756.0, T2: 5182.0}
+        """
+    _n_data = [3, 4]
```

Comment:
The `_n_data` variable is pretty inscrutable

Reply:
It's the list of valid falloff coefficient array lengths. I added a comment where it's first defined (`FalloffRateTests`).